### ISSUE TITLE
LitElement + Ckeditor5 classic

### ISSUE THREAD

**Issue by aadamsx** [Polymer/lit-element#239]
I can only get the new Ckeditor5 classic to render properly OUTSIDE a lit-html render.  The editor inside the render does not get rendered properly.

I tried fooling around with firstUpdated and udpated, but nothing worked.

https://stackblitz.com/edit/lit-element-example-3r3dva?file=index.js

**Comment by aadamsx** [comment]
> Try this.querySelector( '#editor' ) instead of document.querySelector( '#editor' )

Doesn't work.  I updated the stackblitz project accordingly :)

**Comment by web-padawan** [comment]
Appears to be submitted as ckeditor/ckeditor5-engine#692

**Comment by aadamsx** [comment]
Looks like shadydom needs changes as well? https://github.com/webcomponents/shadydom/issues/113

**Comment by web-padawan** [comment]
@aadamsx yep, this should be polyfilled at some point, but first we should see the consensus across implementors (Google, Apple and Mozilla). In the meanwhile, I recommend closing this issue in favour of the `shadydom` issue mentioned above, it would make more sense to continue discussion there.

**Comment by aadamsx** [comment]
@web-padawan I'd like to track it here too if you don't mind?  Won't LitElement need to make changes too?

**Comment by web-padawan** [comment]
@aadamsx no it would not need any cganges. Basically there are 2 things needed:
1. Polyfill `getSelection` on ShadowRoot, which has to be done in shadydom
2. Update CKEditor to use `getRootNode().getSelection()` in case if root node is not a document

**Comment by aadamsx** [comment]
@web-padawan so it sounds like this issue is going to be around a LONG time...

**Comment by blikblum** [comment]
I used rendeRoot `this.renderRoot.querySelector( '#editor' )` and the editor was rendered but seems that there's another issue probably the cited above

**Comment by aadamsx** [comment]
Could you fork the example and add in your code so we can see what you’re taking about @blikblum?

**Comment by justinfagnani** [comment]
@aadamsx I couldn't load your StackBlitz link

**Comment by aadamsx** [comment]
I just updated the link: https://stackblitz.com/edit/lit-element-ckeditor5

**Comment by blikblum** [comment]
https://stackblitz.com/edit/lit-element-ckeditor5-3vcntj?file=index.js

document.querySelector does not cross shadowDom boundaries so you need to query from shadowRoot

**Comment by aadamsx** [comment]
@blikblum yeah, the render is not correct.  I hope this issue gets resolved quickly, as I will not be able to fully use LitElement without a proper editor.

**Comment by blikblum** [comment]
@aadamsx you can disable shadow dom with createRenderRoot. I updated the example in stackblitz

**Comment by aadamsx** [comment]
Thanks @blikblum ... so the createRenderRoot steps outside the shadow dom for the entire class?

The docs say:

> Implement to customize where the element's template is rendered by returning an element into which to render. By default this creates a shadowRoot for the element. To render into the element's childNodes, return this.

This sounds like a forgein language to me.  I wonder why it's not written more planly to make it more friendly to new compers?

> Implement to customize where the element's template is rendered 

What does this mean, to render the element "outside" the shadow dom?  Can the element be render anywhere else, what are the options?

> by returning an element into which to render.

By returning an element into which to render??  Man this is confusing...  The docs can really be more user friendly for guys like me. :)

> By default this creates a shadowRoot for the element. 

But I thought it allows to render outside the showdow dom... but it creates a shadow root, what's that?

> To render into the element's childNodes, return this.

Ah, this is what you example did, returned ```this```, which rendered into a childNode?  What's a child node?

**Comment by thepassle** [comment]
@aadamsx https://polymer.slack.com/archives/CATC4MEUS/p1538772989000100

Overwriting `createRenderRoot()` will cause the template to be rendered to lightdom (had to check myself to make sure though).

**Comment by aadamsx** [comment]
Using the Light Dom worked.  But in order to use the ckeditor inside another template, like wouild be the case in more normal situations, it seems even the "parent" LitElement class needs ```createRenderRoot()``` (rendered to Light-Dom).  

https://stackblitz.com/edit/lit-element-ckeditor5-create-render-root

**Comment by web-padawan** [comment]
See my comment here: https://github.com/Polymer/lit-element/issues/245#issuecomment-427601365

The problem is, whenever you pick some JS library, open the "issues" and search for "Shadow DOM" there. This mostly affects WYSIWYG editors because of `getSelection` API but there are also other DOM API, e.g. `activeElement` which also should be obtained from shadow root, when it is used.
